qcoro/qml/qcoroqmltask.h

```diff
@@ -68,6 +86,19 @@ struct QCOROQML_EXPORT QmlTask {
     Q_INVOKABLE void then(QJSValue func);
 
 private:
+    template <typename Awaitable>
+    static QCoro::Task<QVariant> toTask(Awaitable &&future) {
+        using AwaiterType = typename QCoro::detail::awaiter_type_t<std::remove_cvref_t<Awaitable>>;
+        using ReturnType = decltype(AwaiterType(future).await_resume());
```

Comment:
Despite its name the `awaiter_type_t` is not a generic type - it's only defined for the Qt wrappers in QCoro. So if `future` is a generic awaitable type, this will fail.

However, you can obtain the final awaiter type as a result of the `await_transform()` overload:

```cpp
using AwaiterType = decltype(std::declval<QCoro::TaskPromiseBase>().await_transform(std::forward<Awaitable>(future)));
```

Equally, the `ReturnType` will not work for types that have `operator co_await()`. You can write some type trait for this, something like this (untested):

```cpp
template<typename T>
struct awaitable_return_type {
  using type = decltype(std::declval<T>().await_resume());
};

template<QCoro::detail::has_operator_coawait T>
struct awaitable_return_type<T> {
    using type = typename awaitable_return_type<decltype(std::declval<T>().operator co_await())>::type;
};

...

using ReturnType = awaitable_return_type<AwaiterType>();
```